internal/pkg/manifest/workload_applyenv_test.go

```diff
@@ -0,0 +1,1846 @@
+					Timeout: &mockTimeoutTest,
+				}
+			},
+			wanted: func(svc *LoadBalancedWebService) {
+				mockTimeoutTest := 0 * time.Second
+				svc.ImageConfig.HealthCheck = &ContainerHealthCheck{
+					Timeout: &mockTimeoutTest,
+				}
+			},
+		},
+		"timeout not overridden": {
+			inSvc: func(svc *LoadBalancedWebService) {
+				mockTimeout := 60 * time.Second
+				svc.ImageConfig.HealthCheck = &ContainerHealthCheck{
+					Timeout: &mockTimeout,
+				}
+				svc.Environments["test"].ImageConfig.HealthCheck = &ContainerHealthCheck{}
+			},
+			wanted: func(svc *LoadBalancedWebService) {
+				mockTimeoutTest := 60 * time.Second
+				svc.ImageConfig.HealthCheck = &ContainerHealthCheck{
+					Timeout: &mockTimeoutTest,
+				}
+			},
+		},
+		"start_period overridden": {
+			inSvc: func(svc *LoadBalancedWebService) {
+				mockStartPeriod := 10 * time.Second
+				mockStartPeriodTest := 300 * time.Second
+				svc.ImageConfig.HealthCheck = &ContainerHealthCheck{
+					StartPeriod: &mockStartPeriod,
+				}
+				svc.Environments["test"].ImageConfig.HealthCheck = &ContainerHealthCheck{
+					StartPeriod: &mockStartPeriodTest,
+				}
+			},
+			wanted: func(svc *LoadBalancedWebService) {
+				mockStartPeriod := 300 * time.Second
+				svc.ImageConfig.HealthCheck = &ContainerHealthCheck{
+					StartPeriod: &mockStartPeriod,
+				}
+			},
+		},
+		"start_period explicitly overridden by empty value": {
+			inSvc: func(svc *LoadBalancedWebService) {
+				mockStartPeriod := 10 * time.Second
+				mockStartPeriodTest := 0 * time.Second
+				svc.ImageConfig.HealthCheck = &ContainerHealthCheck{
+					StartPeriod: &mockStartPeriod,
+				}
+				svc.Environments["test"].ImageConfig.HealthCheck = &ContainerHealthCheck{
+					StartPeriod: &mockStartPeriodTest,
+				}
+			},
+			wanted: func(svc *LoadBalancedWebService) {
+				mockStartPeriod := 0 * time.Second
+				svc.ImageConfig.HealthCheck = &ContainerHealthCheck{
+					StartPeriod: &mockStartPeriod,
+				}
+			},
+		},
+		"start_period not overridden": {
+			inSvc: func(svc *LoadBalancedWebService) {
+				mockStartPeriod := 10 * time.Second
+				svc.ImageConfig.HealthCheck = &ContainerHealthCheck{
+					StartPeriod: &mockStartPeriod,
+				}
+				svc.Environments["test"].ImageConfig.HealthCheck = &ContainerHealthCheck{}
+			},
+			wanted: func(svc *LoadBalancedWebService) {
+				mockStartPeriod := 10 * time.Second
+				svc.ImageConfig.HealthCheck = &ContainerHealthCheck{
+					StartPeriod: &mockStartPeriod,
+				}
+			},
+		},
+	}
+
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			var inSvc, wantedSvc LoadBalancedWebService
+			inSvc.Environments = map[string]*LoadBalancedWebServiceConfig{
+				"test": {},
+			}
+
+			tc.inSvc(&inSvc)
+			tc.wanted(&wantedSvc)
+
+			got, err := inSvc.ApplyEnv("test")
+
+			require.NoError(t, err)
+			require.Equal(t, &wantedSvc, got)
+		})
+	}
+}
+
+func Test_ApplyEnv_Entrypoint(t *testing.T) {
```

Comment:
What should be the guideline to our team for when to create a test?
 
We should aim at:
1. Ensuring each field is covered.
2. There should be only one test function where a field is tested. We don't want to make it painful to update the tests.

```
If you have have a multi-type field (such as string or string slice)
   -> Then that should get its own test.
  
Else if you have a struct field (composed of multiple simple types)
   -> Then it should get its own test.
   
Otherwise (a field is just a simple type like "string" or "*string")
   -> The field should be covered in the parent field's test.
```

Reply:
Here is a "guideline" that I came up with:

When writing tests for a field `F` (e.g. `TestApplyEnv_Image`)

* For each sub field `f`:
    * If `f` has subfields, or if `f` is a composite field: write two test cases for this `f` inside e.g. `TestApplyEnv_Image`:( `f` is non-nil,  `f` is nil)
    * Otherwise, write three test cases for `f`: (nil, non-nil and empty value, non-nil and non-empty value)
* For each sub field `f`:
    * If `f` is mutually exclusive with another field (e.g. it’s possible that `F` is a composite filed):
        * Add test cases that check they are mutually exclusive (one is set nil when the other is not)
* For each sub field `f`:
    * If `f` has subfields, or if `f` is a composite field: write another test group for this field (`TestApplyEnv_Image_Build`)